### PR TITLE
fix(runtime): redact absolute runtime context paths

### LINKED ISSUE #1279: [Bug]: Runtime reports leak absolute parent module paths
## Summary

Codex Security reported a **low** finding in the current implementation:

Runtime tracing hooks record parent filenames and entrypoint URLs as absolute filesystem paths or file:// URLs. This commit adds Parent and Entrypoint fields to runtime events, stores them with only file:// stripping and slash normalization, and then copies them into report.RuntimeUsage.ParentModules and Entrypoints. Those fields are serialized in JSON reports and are also emitted in table/TUI and SARIF output.

Full finding: https://chatgpt.com/codex/cloud/security/findings/2a6a729e263c8191a6ba3686da421036

## Affected paths

- `scripts/runtime/require-hook.cjs`
- `scripts/runtime/loader.mjs`
- `internal/runtime/trace_types.go`
- `internal/runtime/trace_load.go`
- `internal/runtime/trace_annotate.go`
- `internal/report/model/dependency.go`
- `internal/report/sarif.go`
- `internal/report/format_table_values.go`

## Reconciliation

- Finding ID: `2a6a729e263c8191a6ba3686da421036`
- Detected: `2026-07-13T13:17:03.301484Z`
- Origin commit: `0faf0792fee5e210c9a64f4d45b70682a53e728d`
- Current `main` reviewed at: `8135de65e2948f1f0571dc0f769505cf2af0d690`
- Result: no merged post-detection fix or existing issue was found for this finding.

## Acceptance criteria

- [ ] Add a focused regression test that fails on the reported behavior.
- [ ] Correct the behavior without weakening existing trust, compatibility, or output contracts.
- [ ] Prove traversal and symlink cases remain confined to the intended trusted root.
- [ ] Run the targeted package/workflow tests and the repository CI gate.

### PR DESCRIPTION
**Pull request by ben-ranford**
## Summary

Runtime parent-module and entrypoint metadata could expose absolute host paths through JSON, table, TUI, and SARIF reports.

This replacement for superseded #1416 stays within #1279: report-visible runtime context now preserves useful repo-relative identity while deterministically redacting paths outside the repository.

Closes #1279
Supersedes #1416

## Root cause

Runtime annotation copied normalized parent and entrypoint strings directly into the report model. Trace loading also removed lowercase `file://` syntax without preserving URL authority, so server and localhost paths could be mistaken for repo-relative values before they reached report renderers.

## Changes

- Render in-repository parent modules and entrypoints as repo-relative paths.
- Render external, traversal, Windows, UNC, and external `file://` paths as `external:<basename>`.
- Reserve and escape `external:` and `literal:` prefixes so redacted paths cannot collide with genuine visible values.
- Preserve non-path module identifiers such as `node:fs` and `@scope/pkg`.
- Merge counts only after deterministic visibility mapping, while keeping repo and external identities distinct.
- Preserve identity when the configured repository root is itself a symlink.
- Preserve and redact `file://` authority and UNC syntax before repository containment so server/share paths stay external.
- Add focused regressions for absolute paths, traversal, Windows/file URLs, symlinked roots, reserved-prefix collisions, and edge-case normalization.

## Validation

```bash
go test ./internal/runtime -run 'TestAnnotateRuntimeContext|TestLoadAndAnnotateRuntimeContextRedactsFileURLAuthorities' -count=1
go test ./internal/analysis -run '^TestServiceAnalysePythonRuntimeTraceIntegration$' -count=1
make ci
```

- Scoped diff: 4 files, 380 additions, 9 deletions.
- New-code duplication: 0.00%.
- Aggregate coverage: 98.3%; `internal/runtime` coverage: 98.3%.
- Security scan: 0 issues.
- Vulnerability scan: 0 vulnerabilities.
- Memory benchmark gate: passed.
- Independent scoped review: approved after collision-proof namespace tests.

Regression-Test: ./internal/runtime::TestAnnotateRuntimeContextRedactsUnsafeVisiblePaths
Regression-Test: ./internal/runtime::TestAnnotateRuntimeContextPreservesIdentityThroughSymlinkedRepoRoot
Regression-Test: ./internal/runtime::TestLoadAndAnnotateRuntimeContextRedactsFileURLAuthorities

## Risk and compatibility

- Breaking changes: report-visible absolute runtime context paths are intentionally replaced with repo-relative or redacted labels.
- Migration required: consumers matching raw absolute `parentModules` or `entrypoints` values must use the new stable labels.
- Performance impact: one bounded path-normalization pass over the existing parent/entrypoint count maps.
- Memory benchmark impact: none; the full memory benchmark gate passed.
- Runtime impact: capture remains unchanged; trace loading now preserves file-URL authority for safe downstream redaction.

## Checklist

- [x] Tests added/updated for behavior changes
- [x] Docs updated (README/docs/schema) if needed
- [x] `memory-approved` requested/applied if intentional memory benchmark regressions exceed CI thresholds
- [x] No unrelated changes included
- [x] Ready for review

### CODE REVIEW

**Comment by ben-ranford** [comment]
@codex review latest commit 91d2bed608ce5692b435e73c870b95975475cae6

**Comment by ben-ranford** [comment]
@codex review latest commit 426ab0e02cbd93d2e1570a5ebc2c9c7ad46d200e

File: internal/runtime/trace_annotate.go [hunk 1]
```diff
@@ -209,6 +216,147 @@ func runtimeModules(values map[string]int) []report.RuntimeModuleUsage {
...
+		return "", false
+	}
+	return filepath.ToSlash(relative), true
+}
+
+func normalizeRuntimeContextValue(value string) (string, bool) {
+	value = strings.ReplaceAll(value, `\`, "/")
+	fileURL := len(value) >= len(fileURLPrefix) && strings.EqualFold(value[:len(fileURLPrefix)], fileURLPrefix)
+	if fileURL {
+		value = value[len(fileURLPrefix):]
```
[the remark this answers is not in the document]
Comment (ben-ranford) [reply internal/runtime/trace_annotate.go:325]: Fixed in 1cc3f2f6d1fd6ffbb3a44e986862ebc9e30408dd

**Comment by ben-ranford** [comment]
@codex review latest commit 1cc3f2f6d1fd6ffbb3a44e986862ebc9e30408dd

File: internal/runtime/trace_annotate.go [hunk 2]
```diff
@@ -209,6 +216,147 @@ func runtimeModules(values map[string]int) []report.RuntimeModuleUsage {
...
+	if !fileURL && strings.Contains(value, "://") && !runtimeContextWindowsDrivePath(value) {
+		return visibleRuntimeContextLiteral(value)
+	}
+	if !fileURL && !runtimeContextLooksLikePath(value) {
+		return visibleRuntimeContextLiteral(value)
+	}
+	uncPath := strings.HasPrefix(value, "//")
+	value = path.Clean(value)
+	if repoRelative, ok := runtimeContextRepoRelativePath(repoPath, value); ok {
+		return visibleRuntimeContextLiteral(repoRelative)
```
[the remark this answers is not in the document]
Comment (ben-ranford) [reply internal/runtime/trace_annotate.go:258]: Fixed in ba5712e82ec3d54ead81d0fe621ceaa91158bfb9

**Comment by ben-ranford** [comment]
@codex review latest commit ba5712e82ec3d54ead81d0fe621ceaa91158bfb9

File: internal/runtime/trace_annotate.go [hunk 3]
```diff
@@ -209,6 +216,154 @@ func runtimeModules(values map[string]int) []report.RuntimeModuleUsage {
...
```
[the remark this answers is not in the document]
Comment (ben-ranford) [reply internal/runtime/trace_annotate.go:258]: Fixed in c09ee3180dda8355322f18a13dd0b00453620212

**Comment by ben-ranford** [comment]
@codex review latest commit c09ee3180dda8355322f18a13dd0b00453620212
